src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -0,0 +1,269 @@
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <BeastConfig.h>
+#include <ripple/app/misc/ValidatorList.h>
+#include <ripple/basics/Slice.h>
+#include <ripple/basics/StringUtilities.h>
+#include <ripple/core/ConfigSections.h>
+#include <boost/regex.hpp>
+#include <fstream>
+
+namespace ripple {
+
+static
+PublicKey
+asPublicKey(RippleAddress const& raPublicKey)
+{
+    auto const& blob = raPublicKey.getNodePublic();
+
+    if (blob.empty())
+        LogicError ("Can't convert invalid RippleAddress to PublicKey");
+
+    return PublicKey(Slice(blob.data(), blob.size()));
+}
+
+ValidatorList::ValidatorList (beast::Journal j)
+    : j_ (j)
+{
+}
+
+boost::optional<std::string>
+ValidatorList::member (RippleAddress const& identity) const
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
+    auto const publicKey = asPublicKey (identity);
+
+    auto ret = ephemeral_.find (publicKey);
+
+    if (ret != ephemeral_.end())
+        return ret->second;
+
+    ret = permanent_.find (publicKey);
+
+    if (ret != permanent_.end())
+        return ret->second;
+
+    return boost::none;
+}
+
+bool
+ValidatorList::trusted (RippleAddress const& identity) const
+{
+    return static_cast<bool> (member(identity));
+}
+
+bool
+ValidatorList::insertEphemeralKey (
+    PublicKey const& identity,
+    std::string comment)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
+    if (permanent_.find (identity) != permanent_.end())
+    {
+        JLOG (j_.error) <<
+            toBase58 (TokenType::TOKEN_NODE_PUBLIC, identity) <<
+            ": ephemeral key exists in permanent table!";
+        return false;
+    }
+
+    return ephemeral_.emplace (identity, comment).second;
+}
+
+bool
+ValidatorList::removeEphemeralKey (
+    PublicKey const& identity)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+    return ephemeral_.erase (identity);
+}
+
+bool
+ValidatorList::insertPermanentKey (
+    RippleAddress const& identity,
+    std::string comment)
+{
+    std::lock_guard <std::mutex> sl (mutex_);
+
+    auto const publicKey = asPublicKey (identity);
+
+    if (ephemeral_.find (publicKey) != ephemeral_.end())
+    {
+        JLOG (j_.error) <<
+            toBase58 (TokenType::TOKEN_NODE_PUBLIC, publicKey) <<
+            ": permanent key exists in ephemeral table!";
+        return false;
+    }
+
+    return permanent_.emplace (publicKey, comment).second;
```

Comment:
Same as above: `std::move(comment)` or pass the parameter in by `const &` instead of value.